### PR TITLE
Adds Miasma Organ Rotting

### PR DESCRIPTION
**Pull request by BurgerLUA**
## About The Pull Request

Organs that rot now create miasma. 

Note that miasma emissions will stop once the organ is fully decayed, and the amount decayed per second is based on the damage dealt to organs.

Roughly the amount of miasma created from one fully-decayed corpse will be about 30 moles of miasma. For reference, a 1x1 room contains 104 moles of oxygen+nitrogen. During testing, I murdered and left Pun Pun to rot for a bit in a 4x3 backroom. Miasma levels were about 2.2% until they were about 70% rotted.

## Why It's Good For The Game

Miasma is hardly used as a gas in gameplay, and needs to honestly exist in more places. Also, there needs to be a way to locate corpses easier that are stuffed in lockers or in small rooms to prevent soft round removal where a body is left in a place for ages and no one notices.

With the nerf to miasma in #649, the effects of miasma exposure shouldn't be as bad as it was before.

## Changelog

:cl: BurgerBB
add: Organs that rot now create miasma. Note that miasma emissions will stop once the organ is fully decayed.
/:cl:

### CODE REVIEW

**Comment by Majkl-J** [comment]
Should probably be on hold until this https://github.com/tgstation/tgstation/pull/79687

Then we can probably make organs handle it instead

**Comment by BurgerLUA** [comment]
> Should probably be on hold until this [tgstation/tgstation#79687](https://github.com/tgstation/tgstation/pull/79687)
> 
> Then we can probably make organs handle it instead

I specifically avoided making organs handle it because having 6-8 organs per person per life tick each calling atmos creation procs is insane

File: code/modules/mob/living/carbon/life.dm [hunk 1]
```diff
@@ -490,13 +490,23 @@
...
 			if(stat != DEAD)
 				break
+		//BUBBERSTATION CHANGE START: MIASMA ORGAN ROT
+		if(rot_count > 0) //This is going to be weird if there are mechanics that cause an organ to heal if you're dead, but at least this saves performance.
+			var/turf/T = get_turf(src) //We don't check the loc because of lockers. Stasis bodybags would prevent this from running anyways.
+			if(isopenturf(T) && !isspaceturf(T)) //Only spawn miasma on floor turfs and not in space turfs.
+				var/turf/open/OT = T
+				if(!OT.planetary_atmos) //Don't spawn miasma when there is open air to the sky.
+					var/miasma_to_spawn = (rot_count/(6*100))*30 //There are about 6 rottable organs in each mob, all with 100 health. If a person is 100% rotted, they should spawn 30 moles of miasma. (A 1x1 tile has 104 moles of oxygen+nitrogen).
+					OT.atmos_spawn_air("[GAS_MIASMA]=[miasma_to_spawn];[TURF_TEMPERATURE(src.bodytemperature)]")
```
Comment (nevimer) [inline code/modules/mob/living/carbon/life.dm:508]: Organs outside of body??

Comment (BurgerLUA) [reply code/modules/mob/living/carbon/life.dm:508]: Organs outside body do create miasma. This is handled by the rot component which is not touched in this PR.

Comment (nevimer) [reply code/modules/mob/living/carbon/life.dm:508]: I don't think they actually do.

Comment (BurgerLUA) [reply code/modules/mob/living/carbon/life.dm:508]: Oops, typo. They do not create miasma, but they are handled by the rot component.

**Comment by theselfish** [comment]
xa3qmcdps2ia1

**Comment by Majkl-J** [comment]
> > Should probably be on hold until this [tgstation/tgstation#79687](https://github.com/tgstation/tgstation/pull/79687)
> > Then we can probably make organs handle it instead
> 
> I specifically avoided making organs handle it because having 6-8 organs per person per life tick each calling atmos creation procs is insane

Fair enough.

**Comment by Detective-Google** [comment]
miasma was intentionally removed from generation by corpses for being totally miserable so I don't think we should add that system back indirectly.

**Comment by yorii** [comment]
I think it's great, you can use a rotting room full of corpses to create oxygen for the station...

File: code/modules/surgery/organs/internal/_internal_organ.dm [hunk 2]
```diff
@@ -43,7 +43,7 @@
 /obj/item/organ/internal/on_death(seconds_per_tick, times_fired) //runs decay when outside of a person
 	if(organ_flags & (ORGAN_ROBOTIC | ORGAN_FROZEN))
 		return
-	apply_organ_damage(decay_factor * maxHealth * seconds_per_tick)
+	return -apply_organ_damage(decay_factor * maxHealth * seconds_per_tick) //BUBBERSTATION CHANGE: ADDS RETURN
```
[the remark this answers is not in the document]
Comment (Majkl-J) [reply code/modules/surgery/organs/internal/_internal_organ.dm:46]: I am befuddled with what you're trying to do here

File: code/modules/mob/living/carbon/life.dm [hunk 3]
```diff
@@ -490,13 +490,23 @@
...
-				organ.on_death(seconds_per_tick, times_fired)
+				rot_count += organ.on_death(seconds_per_tick, times_fired) //BUBBERSTATION ADDITION: MIASMA_COUNT. NOTE THIS ISN'T CALLED ON ROBOTIC ORGANS.
 			// We need to re-check the stat every organ, as one of our others may have revived us
 			if(stat != DEAD)
 				break
+		//BUBBERSTATION CHANGE START: MIASMA ORGAN ROT
+		if(rot_count > 0) //This is going to be weird if there are mechanics that cause an organ to heal if you're dead, but at least this saves performance.
+			var/turf/T = get_turf(src) //We don't check the loc because of lockers. Stasis bodybags would prevent this from running anyways.
+			if(isopenturf(T) && !isspaceturf(T)) //Only spawn miasma on floor turfs and not in space turfs.
+				var/turf/open/OT = T
```
[the remark this answers is not in the document]
Comment (Majkl-J) [reply code/modules/mob/living/carbon/life.dm:505]: image

File: code/modules/surgery/organs/internal/_internal_organ.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (BurgerLUA) [reply code/modules/surgery/organs/internal/_internal_organ.dm:46]: It needs to be negative. Apply organ damage returns a negative number if an organ was damaged.

**Pull request merged**